File: backend/internal/services/recipe_service.go
package services

import (
	"errors"
	"fmt"

	"github.com/yourusername/scentora-backend/internal/models"
	"github.com/yourusername/scentora-backend/internal/repository"
)

type RecipeService struct {
	recipeRepo     *repository.RecipeRepository
	versionRepo    *repository.RecipeVersionRepository
	ingredientRepo *repository.RecipeIngredientRepository
	noteRepo       *repository.RecipeNoteRepository
	tagRepo        *repository.RecipeTagRepository
	accordRepo     *repository.AccordRepository
}

func NewRecipeService(
	recipeRepo *repository.RecipeRepository,
	versionRepo *repository.RecipeVersionRepository,
	ingredientRepo *repository.RecipeIngredientRepository,
	noteRepo *repository.RecipeNoteRepository,
	tagRepo *repository.RecipeTagRepository,
	accordRepo *repository.AccordRepository,
) *RecipeService {
	return &RecipeService{
		recipeRepo:     recipeRepo,
		versionRepo:    versionRepo,
		ingredientRepo: ingredientRepo,
		noteRepo:       noteRepo,
		tagRepo:        tagRepo,
		accordRepo:     accordRepo,
	}
}

// CreateRecipe creates a new recipe with validation
func (s *RecipeService) CreateRecipe(userID string, req *models.CreateRecipeRequest) (*models.Recipe, error) {
	// Validate input
	if req.Name == "" {
		return nil, errors.New("recipe name is required")
	}

	// Check for duplicate name
	exists, err := s.recipeRepo.Exists(userID, req.Name)
	if err != nil {
		return nil, fmt.Errorf("failed to check recipe existence: %w", err)
	}
	if exists {
		return nil, errors.New("recipe with this name already exists")
	}

	// Set default status if not provided
	status := "draft"
	if req.Status != nil {
		status = *req.Status
	}

	// Create recipe
	recipe := &models.Recipe{
		UserID:         userID,
		Name:           req.Name,
		Description:    req.Description,
		TargetVolumeMl: req.TargetVolumeMl,
		Status:         status,
	}
	err = s.recipeRepo.Create(recipe)
	if err != nil {
		return nil, fmt.Errorf("failed to create recipe: %w", err)
	}

	return recipe, nil
}

// GetRecipe retrieves a recipe by ID with all related data
func (s *RecipeService) GetRecipe(recipeID, userID string) (*models.RecipeResponse, error) {
	// Get recipe
	recipe, err := s.recipeRepo.FindByID(recipeID, userID)
	if err != nil {
		return nil, fmt.Errorf("recipe not found: %w", err)
	}

	// Get tags
	tags, err := s.tagRepo.FindByRecipeID(recipeID)
	if err != nil {
		return nil, fmt.Errorf("failed to load tags: %w", err)
	}

	// Get active version
	var activeVersion *models.RecipeVersionResponse
	if recipe.ActiveVersionID != nil {
		version, err := s.versionRepo.FindByID(*recipe.ActiveVersionID)
		if err == nil {
			activeVersion = &models.RecipeVersionResponse{
				ID:            version.ID,
				RecipeID:      version.RecipeID,
				VersionNumber: version.VersionNumber,
				Name:          version.Name,
				Notes:         version.Notes,
				IsActive:      version.IsActive,
				CreatedAt:     version.CreatedAt,
			}
		}
	}

	// Build response
	response := &models.RecipeResponse{
		ID:              recipe.ID,
		UserID:          recipe.UserID,
		Name:            recipe.Name,
		Description:     recipe.Description,
		TargetVolumeMl:  recipe.TargetVolumeMl,
		Status:          recipe.Status,
		ActiveVersionID: recipe.ActiveVersionID,
		ActiveVersion:   activeVersion,
		Tags:            tags,
		CreatedAt:       recipe.CreatedAt,
		UpdatedAt:       recipe.UpdatedAt,
	}

	return response, nil
}

// ListRecipes retrieves recipes with filters
func (s *RecipeService) ListRecipes(userID string, status *string, limit, offset int) ([]*models.Recipe, error) {
	var recipes []*models.Recipe
	var err error

	if status != nil && *status != "" {
		recipes, err = s.recipeRepo.FindByStatus(userID, *status)
	} else {
		recipes, err = s.recipeRepo.FindAll(userID)
	}

	if err != nil {
		return nil, fmt.Errorf("failed to list recipes: %w", err)
	}

	return recipes, nil
}

// SearchRecipes searches recipes by query
func (s *RecipeService) SearchRecipes(userID, query string, limit, offset int) ([]*models.Recipe, error) {
	if query == "" {
		return s.ListRecipes(userID, nil, limit, offset)
	}

	recipes, err := s.recipeRepo.Search(userID, query)
	if err != nil {
		return nil, fmt.Errorf("failed to search recipes: %w", err)
	}

	return recipes, nil
}

// UpdateRecipe updates a recipe
func (s *RecipeService) UpdateRecipe(recipeID, userID string, req *models.UpdateRecipeRequest) (*models.Recipe, error) {
	// Get existing recipe
	recipe, err := s.recipeRepo.FindByID(recipeID, userID)
	if err != nil {
		return nil, fmt.Errorf("recipe not found: %w", err)
	}

	// Validate name change
	if req.Name != nil && *req.Name != recipe.Name {
		exists, err := s.recipeRepo.ExistsExcluding(userID, *req.Name, recipeID)
		if err != nil {
			return nil, fmt.Errorf("failed to check recipe existence: %w", err)
		}
		if exists {
			return nil, errors.New("recipe with this name already exists")
		}
		recipe.Name = *req.Name
	}

	// Update fields
	if req.Description != nil {
		recipe.Description = req.Description
	}
	if req.Status != nil {
		// Validate status
		validStatuses := map[string]bool{
			"draft": true, "in_progress": true, "tested": true, "finalized": true, "archived": true,
		}
		if !validStatuses[*req.Status] {
			return nil, errors.New("invalid status")
		}
		recipe.Status = *req.Status
	}

	// Save changes
	err = s.recipeRepo.Update(recipe)
	if err != nil {
		return nil, fmt.Errorf("failed to update recipe: %w", err)
	}

	return recipe, nil
}

// DeleteRecipe deletes a recipe and all related data
func (s *RecipeService) DeleteRecipe(recipeID, userID string) error {
	// Verify recipe exists and belongs to user
	_, err := s.recipeRepo.FindByID(recipeID, userID)
	if err != nil {
		return fmt.Errorf("recipe not found: %w", err)
	}

	// Delete recipe (cascade will handle versions, ingredients, notes, tags)
	err = s.recipeRepo.Delete(recipeID, userID)
	if err != nil {
		return fmt.Errorf("failed to delete recipe: %w", err)
	}

	return nil
}

// GetRecipesByTags retrieves recipes filtered by tags
func (s *RecipeService) GetRecipesByTags(userID string, tags []string, limit, offset int) ([]*models.Recipe, error) {
	if len(tags) == 0 {
		return s.ListRecipes(userID, nil, limit, offset)
	}

	recipes, err := s.recipeRepo.FindByTags(userID, tags)
	if err != nil {
		return nil, fmt.Errorf("failed to find recipes by tags: %w", err)
	}

	return recipes, nil
}

// GetRecipesByCollection retrieves recipes in a collection
func (s *RecipeService) GetRecipesByCollection(userID, collectionID string) ([]*models.Recipe, error) {
	recipes, err := s.recipeRepo.FindByCollection(collectionID, userID)
	if err != nil {
		return nil, fmt.Errorf("failed to find recipes by collection: %w", err)
	}

	return recipes, nil
}

// GetRecipeStats returns statistics for user's recipes
func (s *RecipeService) GetRecipeStats(userID string) (*models.RecipeStats, error) {
	stats := &models.RecipeStats{}

	// Count by status - repo returns map[string]int
	counts, err := s.recipeRepo.CountByStatus(userID)
	if err != nil {
		return nil, fmt.Errorf("failed to count recipes: %w", err)
	}

	stats.DraftCount = counts["draft"]
	stats.InProgressCount = counts["in_progress"]
	stats.TestedCount = counts["tested"]
	stats.FinalizedCount = counts["finalized"]
	stats.ArchivedCount = counts["archived"]
	stats.TotalCount = stats.DraftCount + stats.InProgressCount + stats.TestedCount + stats.FinalizedCount + stats.ArchivedCount

	return stats, nil
}

File: backend/src/services/recipe.service.ts
import { Decimal } from '@prisma/client/runtime/library';
import { prisma } from '../prisma.js';
import { NotFoundError, ValidationError } from '../utils/errors.js';

function recipeToApi(recipe: any) {
  return {
    _id: recipe.id,
    userId: recipe.userId,
    name: recipe.name,
    ...(recipe.description != null ? { description: recipe.description } : {}),
    targetVolumeMl: Number(recipe.targetVolumeMl),
    status: recipe.status,
    ...(recipe.activeVersionId != null ? { activeVersionId: recipe.activeVersionId } : {}),
    createdAt: recipe.createdAt,
    updatedAt: recipe.updatedAt,
  };
}

function recipeResponseToApi(recipe: any, tags: string[], activeVersion: any | null) {
  return {
    _id: recipe.id,
    userId: recipe.userId,
    name: recipe.name,
    ...(recipe.description != null ? { description: recipe.description } : {}),
    targetVolumeMl: Number(recipe.targetVolumeMl),
    status: recipe.status,
    ...(recipe.activeVersionId != null ? { activeVersionId: recipe.activeVersionId } : {}),
    ...(activeVersion != null ? { activeVersion } : {}),
    ...(tags.length > 0 ? { tags } : {}),
    createdAt: recipe.createdAt,
    updatedAt: recipe.updatedAt,
  };
}

export async function createRecipe(userId: string, data: {
  name: string;
  description?: string;
  targetVolumeMl: number;
  status?: string;
}) {
  // Check name uniqueness
  const exists = await prisma.recipe.findFirst({ where: { userId, name: data.name } });
  if (exists) throw new ValidationError('recipe with this name already exists');

  const recipe = await prisma.recipe.create({
    data: {
      userId,
      name: data.name,
      description: data.description ?? null,
      targetVolumeMl: new Decimal(data.targetVolumeMl),
      status: data.status || 'draft',
    },
  });

  return recipeToApi(recipe);
}

export async function getRecipe(recipeId: string, userId: string) {
  const recipe = await prisma.recipe.findFirst({ where: { id: recipeId, userId } });
  if (!recipe) throw new NotFoundError('recipe not found');

  // Get tags
  const tags = await prisma.recipeTag.findMany({ where: { recipeId } });
  const tagNames = tags.map((t) => t.tag);

  // Get active version
  let activeVersion = null;
  if (recipe.activeVersionId) {
    const version = await prisma.recipeVersion.findUnique({ where: { id: recipe.activeVersionId } });
    if (version) {
      activeVersion = {
        _id: version.id,
        recipeId: version.recipeId,
        versionNumber: version.versionNumber,
        name: version.name,
        ...(version.notes != null ? { notes: version.notes } : {}),
        isActive: version.isActive,
        createdAt: version.createdAt,
      };
    }
  }

  return recipeResponseToApi(recipe, tagNames, activeVersion);
}

export async function listRecipes(userId: string, status?: string, limit = 100, offset = 0) {
  const where: any = { userId };
  if (status) where.status = status;

  const recipes = await prisma.recipe.findMany({
    where,
    orderBy: { createdAt: 'desc' },
    take: limit,
    skip: offset,
  });

  return recipes.map(recipeToApi);
}

export async function searchRecipes(userId: string, query: string, limit = 100, offset = 0) {
  const recipes = await prisma.recipe.findMany({
    where: {
      userId,
      OR: [
        { name: { contains: query, mode: 'insensitive' } },
        { description: { contains: query, mode: 'insensitive' } },
      ],
    },
    orderBy: { createdAt: 'desc' },
    take: limit,
    skip: offset,
  });

  return recipes.map(recipeToApi);
}

export async function updateRecipe(recipeId: string, userId: string, data: {
  name?: string;
  description?: string | null;
  targetVolumeMl?: number;
  status?: string;
}) {
  const recipe = await prisma.recipe.findFirst({ where: { id: recipeId, userId } });
  if (!recipe) throw new NotFoundError('recipe not found');

  // Check name uniqueness on rename
  if (data.name && data.name !== recipe.name) {
    const exists = await prisma.recipe.findFirst({
      where: { userId, name: data.name, id: { not: recipeId } },
    });
    if (exists) throw new ValidationError('recipe with this name already exists');
  }

  const updateData: any = {};
  if (data.name !== undefined) updateData.name = data.name;
  if (data.description !== undefined) updateData.description = data.description;
  if (data.targetVolumeMl !== undefined) updateData.targetVolumeMl = new Decimal(data.targetVolumeMl);
  if (data.status !== undefined) updateData.status = data.status;

  const updated = await prisma.recipe.update({ where: { id: recipeId }, data: updateData });
  return recipeToApi(updated);
}

export async function deleteRecipe(recipeId: string, userId: string) {
  const recipe = await prisma.recipe.findFirst({ where: { id: recipeId, userId } });
  if (!recipe) throw new NotFoundError('recipe not found');
  await prisma.recipe.delete({ where: { id: recipeId } });
}
